src/s2e/Plugins/ExecutionTracers/TestCaseGenerator.cpp

```diff
@@ -11,18 +11,64 @@
 #include <iomanip>
 
 #include <boost/regex.hpp>
+#include <klee/Internal/ADT/ImmutableMap.h>
 #include <s2e/Plugins/OSMonitors/Linux/LinuxMonitor.h>
 #include <s2e/Plugins/OSMonitors/Windows/WindowsCrashMonitor.h>
 #include <s2e/S2E.h>
 #include <s2e/S2EExecutionState.h>
 #include <s2e/S2EExecutor.h>
 #include <s2e/Utils.h>
+
 #include "TestCaseGenerator.h"
 
 namespace s2e {
 namespace plugins {
 namespace testcases {
 
+namespace {
+
+class TestCaseGeneratorState : public PluginState {
+private:
+    ConcreteFileTemplates m_concreteFiles;
+
+public:
+    virtual TestCaseGeneratorState *clone() const {
+        return new TestCaseGeneratorState(*this);
+    }
+
+    static PluginState *factory(Plugin *p, S2EExecutionState *s) {
+        return new TestCaseGeneratorState();
+    }
+
+    Data getChunk(const std::string &name) {
+        auto dataPtr = m_concreteFiles.lookup(name);
+        if (dataPtr == nullptr) {
+            return Data();
+        } else {
+            return dataPtr->second;
+        }
+    }
+
+    void addChunk(const std::string &name, unsigned offset, const Data &chunk) {
+        auto data = getChunk(name);
+
+        if (data.size() < (offset + chunk.size())) {
+            data.resize(offset + chunk.size());
+        }
+
+        for (unsigned i = 0; i < chunk.size(); ++i) {
+            data[offset + i] = chunk[i];
+        }
+
+        m_concreteFiles = m_concreteFiles.replace(std::make_pair(name, data));
```

Comment:
Why use the immutable map if you’re essentially mutating it here anyway? Why not just use a standard still map and not worry about some internal Klee data structure?

Reply:
Because STL maps will copy their content to every state. You might get several MB of redundant data in every state. The copy constructor of an immutable map just increases reference counts instead of copying the internals. When you write to it (like here), it does copy on write. And since writes happen usually only in state 0, it saves a lot of space. I guess I can explain that better in the comment.

Reply:
Ok, that is very interesting! This should probably be documented in `docs` in a “tips and tricks for writing plugins” file or something similar. There are probably lots of existing plugins that use `std::map`, so maybe this requires an issue (to document and refactor exisiting plugins if required).

That said, I still feel a bit dirty when I see `internal` outside of Klee code :)

Reply:
Yeah, we might need some refactoring.
The "internal" could be removed. The classes in there are completely independent from KLEE.